### ISSUE TITLE
[data grid] Excel export is broken due to `eval` with `self` `content-security-policy` 

### ISSUE THREAD

**Issue by OlgaKush512** [mui/mui-x#11153]
### Steps to reproduce

Link to live example: https://stackblitz.com/edit/react-wiwppv?file=index.html 

Steps:
1.set content-security-policy header to include script-src: "'self'". Exemple: 
<meta http-equiv="Content-Security-Policy" content="
        upgrade-insecure-requests;
        default-src 'self';
        script-src 'self' ;
        style-src 'unsafe-inline' 'self';
        object-src 'none';
        base-uri 'self';
        connect-src 'self' ;
        font-src data: 'self' ;
        frame-src 'self';
        img-src 'self' data: ;
        manifest-src 'self';
        media-src 'self';
        worker-src blob:;
        frame-ancestors 'self';
        form-action 'self';
    ">
2. try to download Excel

### Current behavior

Message of error :
`Refused to evaluate a string as JavaScript because 'unsafe-eval' is not an allowed source of script in the following Content Security Policy directive: "script-src 'self' ".`

Impossible to download.

### Expected behavior

Downloading a file excel

### Your environment

System:
    OS: macOS 13.3.1
  Binaries:
    Node: 18.17.0 - ~/.nvm/versions/node/v18.17.0/bin/node
    Yarn: Not Found
    npm: 9.6.7 - ~/.nvm/versions/node/v18.17.0/bin/npm
  Browsers:
    Chrome: 119.0.6045.159
    Edge: Not Found
    Safari: 16.4
  npmPackages:
    typescript: 5.2.2 => 5.2.2 

**Search keywords**: export excel datagrid csp

**Comment by cherniavskii** [comment]
@romgrk Can you take a look at this one?

**Comment by romgrk** [comment]
It might be a stackblitz specific error I think? The same code works fine on codesandbox, and our eval checks are wrapped in `try`. 

@OlgaKush512 Can you confirm this is really related to excel export? I see the error in stackblitz right as the frame loads. Can you also confirm if you experience this error in any other situation, like a production build?

This might need to be reported to stackblitz.

**Comment by OlgaKush512** [comment]
Hello! Thank you for responding so quickly.

This error originated on our platform, and it specifically occurs in the production environment. I wrote approximate code to reproduce this error. And yes, on our company's website, we use a datagrid for exporting to Excel. The error occurs in two places: 
1) when we try to use the export button in the toolbar
2) when we separately use the exportDataAsExcel() function.

In other words, when I attempt to use the export feature on our website, an exception is thrown with this error, and we are unable to download the file.

**Comment by romgrk** [comment]
Can you clarify the reproduction steps? Can you also provide stacktraces for the error that you're getting in production?

I've inspected the origin of the error in stackblitz, it doesn't seem to be datagrid code:
![image](https://github.com/mui/mui-x/assets/1423607/6bdc496f-77e3-41c3-829b-220e009613a4)

**Comment by OlgaKush512** [comment]
Hello!
As mentioned earlier, this is an issue on our platform, not just on StackBlitz.

When exporting Excel, we use the function exportDataAsExcel(), and as a result, we encounter an error.

![Capture d’écran 2023-11-27 à 11 53 34](https://github.com/mui/mui-x/assets/92527601/1396f394-8ef3-4358-986e-b5d0f435c951)
![Capture d’écran 2023-11-27 à 11 54 01](https://github.com/mui/mui-x/assets/92527601/a0edcbdc-5957-439b-94fd-b6d35ec60cd7)

**Comment by romgrk** [comment]
So it seems like this is an `eval` inside ExcelJS: https://github.com/exceljs/exceljs/issues/713

@cherniavskii I haven't looked at the export code yet, do you think we could use the bare version of ExcelJS as suggested in the issue above?

**Comment by cherniavskii** [comment]
I've added a comment in the related `exceljs` issue https://github.com/exceljs/exceljs/issues/713#issuecomment-1828759610
If using the bare version of `exceljs` works as a workaround - great, but the ultimate fix should be done on the `exceljs` side.
I'll try to open an issue in `exceljs` with a proposal for issue resolution later.

**Comment by ryancogswell** [comment]
@cherniavskii I see in the exceljs issue you linked to that they believe it was fixed in the most recent release. I'm running into this same problem -- our CSP doesn't allow eval. It would be great if the new exceljs version can be incorporated soon (assuming it does indeed fix the problem).

**Comment by cherniavskii** [comment]
I tried to reproduce the issue again, but using Stackblitz container + vite.
I had to use nonce for Vite dev inline scripts to work, but couldn't reproduce the issue with the data grid or `exceljs` specifically.
I added an inline script with direct `eval()` call to demonstrate that CSP is configured to disallow `eval` calls. But the data grid and Excel export work just fine.

https://stackblitz.com/edit/mrqegsbu

<img width="793" height="264" alt="Image" src="https://github.com/user-attachments/assets/f6e71d32-cbb2-452b-b2d8-d35088ac556e" />

It looks like the reproduction example provided by @OlgaKush512 is related to some kind of Stackblitz issue, not the data grid.
Although on the screenshot from the actual app I see that it's coming from `exceljs` (data grid's dependency).

If this issue still exists, we need a reproduction example.
Otherwise, we won't be able to solve the issue on our end.

**Comment by ryancogswell** [comment]
@cherniavskii The last comment on https://github.com/exceljs/exceljs/issues/713#issuecomment-2451022206 from a year ago indicates they think it was fixed in 4.4.0 which is the exceljs version you are now using, so this is now resolved. We're actually still on a v6 version of the data grid and it even works there with a fresh install because the exceljs version of "^4.3.0" resolves to "4.4.0". I just verified that it works now in our application.

**Comment by cherniavskii** [comment]
@ryancogswell Thanks for the info. I feel like there were multiple CSP-related issues over the lifetime of that issue in the repo, but it's not clear which PRs have fixed them. Maybe `jszip` upgrade, not sure.

> The last comment on https://github.com/exceljs/exceljs/issues/713#issuecomment-2451022206 from a year ago indicates they think it was fixed in 4.4.0

This comment refers to the `.bare.` file, which is not used by default. Is this the one you're using?
Also, the [lack of the lockfile in the exceljs repo](https://github.com/exceljs/exceljs/blob/5bed18b45e824f409b08456b59b87430ded023ab/.gitignore#L56) makes it even harder to track which dependencies are bundled 🫠

**Comment by ryancogswell** [comment]
@cherniavskii I'm not using the `.bare.` file, I'm just using what the data grid pulls in. I'm not clear on what exactly in 4.4.0 fixed it, but I know that last November when I commented on this issue, our lock file showed that we were pulling in exceljs 4.3.0 and at that time our CSP policy caused Excel export to fail. Our CSP policy hasn't changed, but now exceljs 4.4.0 is in our lock file and the Excel export works.
